readme.md
# MLBStats
### The 2018 season is here!
By Jacob Hall

MLBStats is an easy-to-use and powerful JavaScript wrapper for fetching official Major League Baseball stats, scores, and information. MLBStats is in its initial release stages, features and support are continuously being developed.

## Highlighted Features - [Changelog](./changelog.md) / [Upcoming](https://github.com/jakeehall/MLBStats/projects)
* Easy to Use
    * Search functionality
        * Find a game by searching for a team on a date
        * Find all games on a date by searching a date
    * Shortcut to find a date relative to today
    * Flexible Functions, allow for multiple or single searches to be requested using the same function
    * Full documentation and examples
    * Helpful console error messages to resolve bugs
* Full of Functions
    * Look up player(s) stats (batting, and pitching)
    * Look up game(s) stats (linescore, plays, and more...)
    * Look up all games played on a given date
    * Look up bullpens and benched players
    * Support for minor leagues (Coming Soon*)
    * Support for double headers
    * MORE COMING SOON!
* Powerful
    * Written using functional asynchronous code.
        * IT'S FAST!!!
    * Functions are compatible with each other, allowing for the chaining of MLBStats functions.
    * It only takes one request to find every piece of information.
        * Functions return all available data for a given search subject. This saves users from bandwidth lag times when looking up stats of a similar and previous searches.

### Disclaimer
#### PLEASE USE MLBSTATS RESPONSIBLY!
This library and projects using this library, must be used for individual purposes ONLY. To use this library for bulk or commercial purposes you MUST have permission from Major League Baseball (more information below).

>[The accounts, descriptions, data and presentation in the referring page (the "Materials") are proprietary content of MLB Advanced Media, L.P ("MLBAM"). Only individual, non-commercial, non-bulk use of the Materials is permitted and any other use of the Materials is prohibited without prior written authorization from MLBAM. Authorized users of the Materials are prohibited from using the Materials in any commercial manner other than as expressly authorized by MLBAM.](http://gdx.mlb.com/components/copyright.txt)

### Dependency Credits
* #### [fetch (Polyfill)](https://github.com/github/fetch) - GitHub
* #### [whatwg-fetch](https://github.com/matthew-andrews/fetch) - Matt Andrews
* #### [xml2json](http://goessner.net/download/prj/jsonxml/) - Stefan Goessner

## Getting Started
#### Clone or Download the ZIP of _MLBStats.min.js_ from GitHub

The newest version of this file can be found in this directory on GitHub:
```
/min/MLBStats.js
```
**_NOTE: The X's are the version number._**

After downloading the library, place it in your project directory. The only file necessary to include in your project is _MLBStats.min.js_ as the rest of the files are for development of this library only.
#### Include the MLBStats Library in your HTML
```html
<script type="text/javascript" src="MLBStats.X.X.X.min.js"></script>
<!-- Change the X's to match the version number of the library! -->
```
Include this line at the end of your HTML body.

**_NOTE 1: If the MLBStats.min.js file isn't stored in the root directory of the project, change the path in the "src" attribute to where the MLBStats.min.js file is located._**

**_NOTE 2: The MLBStats library MUST be included in the projects HTML above any custom JavaScript or any JavaScript file that uses of the MLBStats library!_**

After the MLBStats library is included in the project, test that it is working by calling **MLBStats.manual()** from the JavaScript console. An alert will popup if the library is working, if there's no popup make sure there's nothing blocking alert popups in your browser, and read the notes above.

#### Alert MLBStats Manual
The MLBStats manual alerts valuable, quick-reference information on how to use the MLBStats library, as well as a links to this readme on GitHub, and other useful references. The manual also displays basic information about the library, such as, contributors and version number in the alert message.

If a problem ever occurs while working with the MLBStats library, call the MLBStats manual function below from the JavaScript console to find help.
``` javascript
MLBStats.manual();
//You should NEVER use this function in production
//This function is ONLY for quick usage help and library info
```

#### Print and Return MLBStats Version
``` javascript
//Return VERSION version
MLBStats.version((main, sub, patch) => {
    console.log(`v${main}.${sub}.${patch}`);
    //OUTPUT: v1.3.0
});
```
This function can be used to check the users library version for compatibility and troubleshooting. For example, if another library used MLBStats as a way of fetching data, that library, could also check to see if the version of MLBStats is compatible with itself.

## Important Objects and Ideas in MLBStats
### Formatted Date Object
> ``` javascript
> let date = {
>     year: 2017,
>     month: 09,
>     day: 16,
>     game: 1, //Used for double-headers
> };
> ```
> **_NOTE: If a single digit month or day is not lead with a zero, a lead zero will be added automatically._**

### Search Object
> ``` javascript
> //EXAMPLE:
> let Search = {
>     gameID: "gid_2017_09_16_sdnmlb_colmlb_1",
>     playerID: 453568,
>     teamID: 115,
>     date: {
>         year: 2017,
>         month: 09,
>         day: 16,
>         game: 1,
>     },
> };
> ```
> ``` javascript
> //Shortcut Day Number or JavaScript Date Object also work in Search
> //Search a players stats for today (Shortcut)
> let Search = {
>     gameID: "gid_2017_09_16_sdnmlb_colmlb_1",
>     playerID: 453568,
>     teamID: 115,
>     date: 0, //SHORTCUT:  -1=Yesterday  0=Today  1=Tomorrow
> };
>
>
> //ADVANCED: Search a players stats for today (JavaScript Date Object)
> let Search = {
>     gameID: "gid_2017_09_16_sdnmlb_colmlb_1",
>     playerID: 453568,
>     teamID: 115,
>     date: new Date(1503980931904), //JS Date Object: 2017-08-28
> };
> ```
> **_NOTE: playerID or teamID, and date, are required to return a gameID using a Search Object. If date is undefined in the Search Object the date will automatically be set to today._**

### rq (Request Query)
> Although rq, the Request Query feature, doesn't seem powerful at first, it allows complicated stat look ups in one step. rq is universal throughout the look up functions in MLBStats, meaning that if a rq Array is created it can be used in all of the look up functions as long as it provides the necessary data in its search for each specific lookup function. rq doesn't only have the ability to pass multiple items at once in a look up function, it also allows for flexibility to provide different types of Search Objects in each element of the rq Array, and the ability to build up a look up queue over time.
>
> Filter and Sort are planned to be added into rq in v2.0.0, but don't worry, rq is forward compatible so code written now will not break as features are added.
>
> ``` javascript
> let rq =
> [{
>     search: {
>         //What were Charlie Blackmon's stats on 2017-09-01?
>         playerID: 453568, //Charlie Blackmon
>         date: {
>             year: 2017,
>             month: 9,
>             day: 1,
>         },
>     },
> },
> {
>     search: {
>         //How did Mark Renyolds do Yesterday?
>         playerID: 448602, //Mark Renyolds
>         date: -1, //SHORTCUT: Yesterday
>     },
> },
> {
>     search: {
>         //Who's playing left field for the Rockies today?
>         teamID: 115,
>         pos: "LF",
>         date: 0, //Today
>     },
> },];
> ```
> **_NOTE: rq (Request Query) can be used to request single or multiple searches at once, in any of the MLBStats look up functions (player, game, bench, gamesOnDate, etc...)._**

### "The Shortcut"
> The shortcut is used to quickly look-up, or automate, days in relation to today. Simply pass an integer value in as a date, and a formatted date will be constructed and used.
> ##### Visualization Key:
> ``` javascript
> date: 1, //Tomorrow
> date: 0, //Today
> date: -1, //Yesterday
> date: -2, //The day before Yesterday
> date: -7, //Last Week on this day
> date: -30, //Last Month
> date: -365, //A Year Ago
> ...
>```

### Formatting a date
> ##### From Shortcut Date, JavaScript Date Object, Formatted Date Object:
> ``` javascript
> //Shortcut:

changelog.md
## v1.3.0 [20180421]
#### Ease-of-Use, Wider Browser/Node Support, Callback/General Speed, and Fixes Update
* ADDED: More pitching stats!
    * Game: bb, er, era, h, hbp, hra, ip, k, np, r, s, whip
    * Season: er, hbp, hra, k, r
* ADDED: A players face image URL is now returned under the 'faceImage' key, when looking up Players.
* ADDED: Support for CommonJS and AMD, while keeping Browser Support.
    * Testing for Node can be found at /examples/nodeapp.js
* ADDED: Browserify + Babel Transpilers
    * All dependencies are now included in one minimized bundle JavaScript file. This allows for dependency support in browsers.
    * There's now a map for the bundle, for easier debugging.
* ADDED: Polyfills for JavaScript fetch and promises.
    * Older browsers are now supported, and Node can now make HTTP requests using the fetch polyfill.
* CHANGED: XMLHttpRequest to fetch
* CHANGED: All functions now return data through callback functions.
    * This change was made to benefit the future of this library. By making all functions have callbacks the API becomes consistent and easier to remember. Also if more async functions are added the library will also immediately become faster.
* CHANGED: Signature of
    * formatDate(date) to formatDate(date, callback)
    * formatGameID(gameID) to formatGameID(gameID, callback)
    * version(log = true) to version(callback)
        * The version function no longer prints the current libraries version to the JavaScript console log. To check your library version, please use the manual() function, or use the version(callback) function to log the version yourself (See example below).
        * The version function now returns the sub-versions through the callback:
            ``` javascript
            //Return VERSION version
            MLBStats.version((main, sub, patch) => {
                console.log(`v${main}.${sub}.${patch}`);
                //OUTPUT: v1.3.0
            });
            ```
* CHANGED: gameToDate function name to gameIDToDate, for consistency.
* CHANGED: A Players date of birth is now returned as a formatted date object.
* CHANGED: A Players height is now returned as a height object.
    ``` javascript
    //A 6 foot, 1 inch player:
    {ft: "6", in: "1"}
    ```
* CHANGED: examples/index.html and readme.md to reflect the changes made in v1.3.0.
* FIXED: Nasty anti-pattern on users side when returning players and fixes issue #1 (Can't lookup the same player multiple times using a single player lookup RQ).
    * Players are no longer returned by their playerID. Instead they are pushed to array in the order that their data is returned from the server.
* FIXED: Player lookup function is now cleaner and faster.
    * cleanPlayerResponse function was removed from the Player lookup function.
    * A new Player object is created rather than modifying the existing one.
* FIXED: RQ would be its own separate item of the returned array of games when using the gamesOnDate lookup.
    * The RQ for gamesOnDate is now stored in each individual game.
* FIXED: Player no longer returns pitching game stats if that player isn't pitching.
* FIXED: test/index.html gamesOnDate example to work with RQ
* FIXED: MLBStats.manual() now logs rather than alerting, this way it can support the Browser and Node.
    * The manual string data is now returned, allowing for the manual to be displayed in different ways, E.g.:
    ``` javascript
        alert(MLBStats.manual()); //Only in browsers
    ```
* FIXED: A warning message will no longer appear in the JavaScript console when a player is looked up.

## v1.2.0 [20171111]
#### Player Update (Continuing through v1.2.X)
* ADDED: Pretty Responses to the player() function:
    ``` javascript
    //For a full list for return attributes, view the readme
    Batting: {Team: Object, Empty: Object, RISP: Object, Loaded: Object, Pitch: {loves: "FT", hates: "FC"}, …}
    Date: {year: 2017, month: "09", day: 24, game: 1}
    Pitching: {Month: Object, Team: Object, Empty: Object, RISP: Object, Loaded: Object, …}
    RQ: {search: Object}
    bats: "R"
    cpos: "P"
    dob: {year: 1995, month: 02, day: 22}
    firstName: "German"
    fullName: "German Marquez"
    height: "6-1"
    lastName: "Marquez"
    number: "48"

src/MLBStats.1.0.0.js
//MLBStats, a Major League Baseball Stats library written in JavaScript
//Created By Jacob Hall

((window) => {
    'use strict';
    function define_library() {
        const version = {main: 1, sub: 0, patch: 0};
        //If you contributed to this project in some way, feel free to include
        //your real name and personal website as an element in the authors array
        const authors = [{name: "Jake Hall", website: "http://jakehall.me"},];

        let MLBStats = {};



        //Alert manual, a quick reference guide for this library
        MLBStats.manual = () => {
            let manual =
                `MLBStats\n` +
                `v${version.main}.${version.sub}.${version.patch}\n\n` +
                `Created By:\n` +
                `${authors[0].name}\t${authors[0].website}\n\n` +
                `For more information on how to use the MLBStats library ` +
                `check out the "readme.md" on GitHub here:\n\n` +
                `(Link coming in next update!)\n\n` +
                `If any issues related to this library occur, please report` +
                `all available issue-related information to our GitHub page!`;
            if (authors.length > 1) {
                manual += `\n\nContributors:\n`;
                for(let i = 1, len = authors.length; i < len; i++) {
                    manual += `${authors[i].name}\t${authors[i].website}\n`;
                }
            }
            alert(manual);
        }



        /*
        Check the users library version for compatibility and troubleshooting
        PASS true, to output log the formatted version string to the console
        RETURN the version object:
        {
            main: 1,
            sub: 0,
            patch: 0,
        }
        */
        MLBStats.version = (log = false) => {
            if(log) {
                console.log(`v${version.main}.${version.sub}.${version.patch}`);
            }
            return version;
        }



        //TRY CATCH THROW
        //PASS valid date formats (listed in errorMsg below)
        //RETURN Formatted Date Object: {year: 2017, month: 9, day: 1, game: 1}
        MLBStats.formatDate = (date) => {
            let errorMsg =
                `Invalid Date Format!\n` +
                `Use one of the lookup formats listed below:\n` +
                `1. Shortcut Date: -1=Yesterday, 0=Today, 1=Tomorrow...\n` +
                `2. Formatted Date Object: {year: 2017, month: 9, day: 1}\n` +
                `3. JavaScript Date Object`;

            //If no date is given, default to today
            if (typeof(date) === 'undefined')
                date = 0;

            //If shortcut is used, create JavaScript Date Object with new date
            //by converting to a JS Date object we can support. Ex: -365
            if (typeof(date) === 'number') {
                const newDate = new Date();
                newDate.setDate(newDate.getDate() + date);
                date = newDate;
            }

            //If date is object (JavaScript Date Object or Formatted Date)
            if (typeof(date) === 'object') {
                let yyyy, mm, dd, g = 1;

                //JavaScript Date Object
                if (date instanceof Date) {
                    yyyy = date.getFullYear();
                    mm = date.getMonth()+1;//January is month 0
                    dd = date.getDate();

                //Formatted Date Object
            } else if ((typeof(date.year) !== 'undefined')
                        && (typeof(date.month) !== 'undefined')
                        && (typeof(date.day) !== 'undefined')) {
                    //Convert new values into numbers in case they were strings
                    yyyy = Number(date.year);
                    mm = Number(date.month);
                    dd = Number(date.day);
                    if (typeof(date.game) !== 'undefined') {
                        g = Number(date.game);
                    }

                //Unknown Format
                } else {
                    console.error(`${errorMsg}\nUnknown Date Object!`);
                    return null;
                }

                //Add leading zero if day or month is a single digit
                if (mm < 10) mm = "0"+mm;
                if (dd < 10) dd = "0"+dd;

                //Return Formatted Date
                return {year: yyyy, month: mm, day: dd, game: g};
            }
            console.error(errorMsg);
            return null;
        }



        //PASS gameID path or gameday
        //RETURN gameID
        MLBStats.formatGameID = (unformattedGameID) => {
            let len = unformattedGameID.length;
            if (len === 26) {
                //2017/09/23/colmlb-sdnmlb-1
                //replace '/' and '-' with '_'
                return 'gid_'+unformattedGameID.replace(/\/|-/g, '_');
            }
        }



        //PASS gameID or Search Object
        //RETURN Formated Date Object
        MLBStats.gameIDtoDate = (search, callback) => {
            MLBStats.searchForGameID(search, (gameID) => {
                //If the gameID is found and that it is a string
                if (gameID !== null && typeof(gameID) === 'string') {

                    //Validate gameID, and extract date and game number
                    if (gameID.search('gid') === 0 && gameID.length === 30) {
                        //EXAMPLE gameID: gid_2017_09_16_sdnmlb_colmlb_1
                        let yyyy = gameID.slice(4, 8);
                        let mm = gameID.slice(9, 11);
                        let dd = gameID.slice(12, 14);
                        let g = gameID.slice(-1);
                        return callback({year: yyyy, month: mm, day: dd, game: g});
                    }

                    //Invalid gameID
                    console.error(`Cannot return Formatted Date from this gameID:\n${gameID}`);
                    return callback(null);

                //Invalid Search Object
                } else {
                    console.error(`Cannot find a gameID with this Search Object:\n${search}`);
                    return callback(null);
                }
            });
        }



        //TRY CATCH THROW
        //PASS gameID as String or Search Object
        //CALLBACK gameID
        MLBStats.searchForGameID = (search, callback) => {
            let errorMsg =
                `Invalid Search Object!\n` +
                `Pass either a gameID as a String, or a Search Object:\n`+
                `{date: {year: 2017, month: 9, day: 1, game: 1}, playerID: 453568}\n` +
                `Note: Must include either a playerID or teamID member!\n`;

            //If gameID is passed in return
            if (typeof(search) === 'string') {
                return callback(search);

            //Else if search object is passed in
            } else if (typeof(search) === 'object') {
                if (typeof(search.gameID) === 'string') {
                    return callback(search.gameID);
                }
                //Format the date passed in, if no date passed set day to today
                search.date = MLBStats.formatDate(search.date);

                //playerID
                if (typeof(search.playerID) === 'number' || typeof(search.playerID) === 'string') {
                    request({
                        leauge: 'mlb',
                        date: search.date,
                        endpoint: `/batters/${search.playerID}_${search.date.game}.xml`,
                    }, (data) => {
                        let gameID = MLBStats.formatGameID(data.batting.game_id);

                        let gameStats = {}, seriesStats = {};

                        //search for series and game stats, set their values if available
                        for (let key in data.batting) {
                            if (!key.startsWith('s_')) {
                                if (key.startsWith('ser_')) {
                                    let keyName = key.slice(4);
                                    seriesStats[keyName] = data.batting[key];
                                } else {
                                    gameStats[key] = data.batting[key];
                                }
                            }
                        }

                        return callback(gameID, gameStats, seriesStats);
                    });

                //teamID
                } else if (typeof(search.teamID) === 'number' || typeof(search.teamID) === 'string') {
                    //games
                    return callback('gid_2017_09_16_sdnmlb_colmlb_1');

                //Either playerID or teamID is required to return a gameID
                } else {
                    console.error(errorMsg);
                    return callback(null);
                }
            //Unrecognized format
            } else {
                console.error(errorMsg);
                return callback(null);
            }
        }



        /*
        PASS rq or an array of rq (Request Query) with search
        {
            search: {
                gameID: formattGameID,
                playerID: playerID,
                teamID: teamID,
                date: date,
            }
        }
        CALLBACK players batting, pitching, and fielding stats
        */
        MLBStats.player = (rq, callback) => {
            //If rq (Request Query) isn't an array make it a single element array
            if (!Array.isArray(rq)) {
                rq = [rq];
            }

            let players = {};
            let itemsProcessed = 0;
            let length = rq.length;

            let addPlayer = (player) => {
                itemsProcessed++;
                if (player !== null) {
                    players[player.batter.id] = player;
                }
                //Only callback after all players info is returned
                if(itemsProcessed === length) {
                    return callback(players);
                }
            }

            //Loop through each player passed
            rq.forEach((item, index, array) => {
                let date = MLBStats.formatDate(item.search.date);

                MLBStats.searchForGameID(item.search, (gameID, gameStats, seriesStats) => {
                    //If no gameID is returned, don't continue executing function
                    if (gameID === null) {
                        addPlayer(null);
                    } else {
                        //Search for batter
                        request({
                            leauge: 'mlb',
                            date: date,
                            endpoint: `/${gameID}/batters/${item.search.playerID}.xml`,
                        }, (batterStats) => {
                            //Create player object, batter and pitcher stored in same player
                            let player = {};

                            //Check if batter request was successful
                            if (batterStats === null) {
                                addPlayer(null);
                            //if successful then format stats
                            } else {
                                player.batter = batterStats.Player;
                                player.batter.game = gameStats;
                                player.batter.series = seriesStats;
                                player.date = date;
                            }

                            //Check if player is a pitcher, if not return now
                            if (batterStats.Player.pos !== 'P') {
                                addPlayer(player);
                            } else {
                                //Search for pitcher
                                request({
                                    leauge: 'mlb',
                                    date: date,
                                    endpoint: `/${gameID}/pitchers/${item.search.playerID}.xml`,
                                }, (pitcherStats) => {
                                    //Check if pitcher request was successful
                                    if (pitcherStats !== null) {
                                        player.pitcher = pitcherStats.Player;
                                    }
                                    addPlayer(player);
                                });
                            }
                        });
                    }
                });
            });
        }



        /*
        PASS rq (Request Query) with search
        REQUIRES date, if no date passed in rq. search then default to today
        {
            search: {
                date: date,
            }
        }
        CALLBACK todays games
        */
        MLBStats.gamesOnDate = (rq, callback) => {
            let date = MLBStats.formatDate(rq.search.date);
            request({
                leauge: 'mlb',
                date: date,
                endpoint: `/master_scoreboard.json`,
            }, (games) => {
                return callback(games.data.games.game);
            });
        }



        //fix pass gameID
        /*
        PASS gameID or rq (Request Query) with search
        {
            search: {
                gameID: formattGameID,
                playerID: playerID,
                teamID: teamID,
                date: date,
            }
        }
        CALLBACK game info
        */
        MLBStats.game = (rq, callback) => {
            //If rq (Request Query) isn't an Array make it single element array
            if (!Array.isArray(rq)) {
                rq = [rq];
            }

            let games = [];
            let itemsProcessed = 0;

            //Loop through each game passed
            rq.forEach((item, index, array) => {
                MLBStats.searchForGameID(item.search, (gameID) => {
                    MLBStats.gameIDtoDate(gameID, (date) => {
                        request({
                            leauge: 'mlb',
                            date: date,
                            endpoint: `/${gameID}/linescore.json`,
                        }, (game) => {
                            games.push(game.data.game);
                            itemsProcessed++;
                            if(itemsProcessed === array.length)
                                return callback(games);
                        });
                    });
                });
            });
        }



        /*
        PASS gameID or rq (Request Query) with search
        {
            search: {
                gameID: formattGameID,
                playerID: playerID,
                teamID: teamID,
                date: date,
            }
        }
        CALLBACK bench
        */
        MLBStats.bench = (rq, callback) => {
            MLBStats.searchForGameID(rq, (gameID) => {
                MLBStats.gameIDtoDate(gameID, (date) => {
                    request({
                        leauge: 'mlb',
                        date: date,
                        endpoint: `/${gameID}/benchO.xml`,
                    }, (bench) => {
                        return callback(bench.bench);
                    });
                });
            });
        }


        return MLBStats;
    }

    //Define library when the page is loaded
    if (typeof(MLBStats) === 'undefined') {
        window.MLBStats = define_library();//create namespace
    }

})(window);


/*
PASS e (endpoint Object)
{
    leauge: 'mlb',
    date: formattedDate,
    endpoint: `/${gameID}/game.xml`,
}
CALLBACK requested object in JSON format
*/
function request(e, callback) {
    const xhr = new XMLHttpRequest();
    xhr.withCredentials = false;
    xhr.addEventListener('readystatechange', function() {
        if (this.readyState === 4) {
            if (this.status === 200) {
                //Check for XML, if XML convert to JSON
                if (e.endpoint.endsWith('xml')) {
                    let xml = new DOMParser().parseFromString(this.responseText, "text/xml");
                    let json = xml2json(xml);
                    //console.warn(json);//DEBUG
                    return callback(JSON.parse(json));
                } else if (e.endpoint.endsWith('json')) {
                    //console.warn(this.responseText);//DEBUG
                    return callback(JSON.parse(this.responseText));
                }
                return callback(null);
            }
        }
    });
    xhr.open('GET', `http://gd2.mlb.com/components/game/` +
                    `${e.leauge}/` +
                    `year_${e.date.year}/` +
                    `month_${e.date.month}/` +
                    `day_${e.date.day}` +
                    `${e.endpoint}`,
                    true);
    xhr.send(null);
}


/*	This work is licensed under Creative Commons GNU LGPL License.
	License: http://creativecommons.org/licenses/LGPL/2.1/
    Version: 0.9
	Author:  Stefan Goessner/2006
	Web:     http://goessner.net/
*/
function xml2json(xml) {
   var X = {
      toObj: function(xml) {
         var o = {};
         if (xml.nodeType==1) {   // element node ..
            if (xml.attributes.length)   // element with attributes  ..
               for (var i=0; i<xml.attributes.length; i++)
                  o[xml.attributes[i].nodeName] = (xml.attributes[i].nodeValue||"").toString();
            if (xml.firstChild) { // element has child nodes ..
               var textChild=0, cdataChild=0, hasElementChild=false;
               for (var n=xml.firstChild; n; n=n.nextSibling) {
                  if (n.nodeType==1) hasElementChild = true;
                  else if (n.nodeType==3 && n.nodeValue.match(/[^ \f\n\r\t\v]/)) textChild++; // non-whitespace text
                  else if (n.nodeType==4) cdataChild++; // cdata section node
               }
               if (hasElementChild) {
                  if (textChild < 2 && cdataChild < 2) { // structured element with evtl. a single text or/and cdata node ..
                     X.removeWhite(xml);
                     for (var n=xml.firstChild; n; n=n.nextSibling) {
                        if (n.nodeType == 3)  // text node
                           o["#text"] = X.escape(n.nodeValue);
                        else if (n.nodeType == 4)  // cdata node
                           o["#cdata"] = X.escape(n.nodeValue);
                        else if (o[n.nodeName]) {  // multiple occurence of element ..
                           if (o[n.nodeName] instanceof Array)
                              o[n.nodeName][o[n.nodeName].length] = X.toObj(n);
                           else
                              o[n.nodeName] = [o[n.nodeName], X.toObj(n)];
                        }
                        else  // first occurence of element..
                           o[n.nodeName] = X.toObj(n);
                     }
                  }
                  else { // mixed content
                     if (!xml.attributes.length)
                        o = X.escape(X.innerXml(xml));
                     else
                        o["#text"] = X.escape(X.innerXml(xml));
                  }
               }
               else if (textChild) { // pure text
                  if (!xml.attributes.length)
                     o = X.escape(X.innerXml(xml));
                  else
                     o["#text"] = X.escape(X.innerXml(xml));
               }
               else if (cdataChild) { // cdata
                  if (cdataChild > 1)
                     o = X.escape(X.innerXml(xml));
                  else
                     for (var n=xml.firstChild; n; n=n.nextSibling)
                        o["#cdata"] = X.escape(n.nodeValue);
               }
            }
            if (!xml.attributes.length && !xml.firstChild) o = null;
         }
         else if (xml.nodeType==9) { // document.node
            o = X.toObj(xml.documentElement);
         }
         else
            console.error("Unhandled node type: " + xml.nodeType);
         return o;
      },
      toJson: function(o, name, ind) {
         var json = name ? ("\""+name+"\"") : "";
         if (o instanceof Array) {
            for (var i=0,n=o.length; i<n; i++)
               o[i] = X.toJson(o[i], "", ind+"\t");
            json += (name?":[":"[") + (o.length > 1 ? ("\n"+ind+"\t"+o.join(",\n"+ind+"\t")+"\n"+ind) : o.join("")) + "]";
         }
         else if (o == null)
            json += (name&&":") + "null";
         else if (typeof(o) == "object") {
            var arr = [];
            for (var m in o)
               arr[arr.length] = X.toJson(o[m], m, ind+"\t");
            json += (name?":{":"{") + (arr.length > 1 ? ("\n"+ind+"\t"+arr.join(",\n"+ind+"\t")+"\n"+ind) : arr.join("")) + "}";
         }
         else if (typeof(o) == "string")
            json += (name&&":") + "\"" + o.toString() + "\"";
         else
            json += (name&&":") + o.toString();
         return json;
      },
      innerXml: function(node) {
         var s = ""
         if ("innerHTML" in node)
            s = node.innerHTML;
         else {
            var asXml = function(n) {
               var s = "";
               if (n.nodeType == 1) {
                  s += "<" + n.nodeName;
                  for (var i=0; i<n.attributes.length;i++)
                     s += " " + n.attributes[i].nodeName + "=\"" + (n.attributes[i].nodeValue||"").toString() + "\"";
                  if (n.firstChild) {
                     s += ">";
                     for (var c=n.firstChild; c; c=c.nextSibling)
                        s += asXml(c);
                     s += "</"+n.nodeName+">";
                  }
                  else
                     s += "/>";
               }
               else if (n.nodeType == 3)
                  s += n.nodeValue;
               else if (n.nodeType == 4)
                  s += "<![CDATA[" + n.nodeValue + "]]>";
               return s;
            };
            for (var c=node.firstChild; c; c=c.nextSibling)
               s += asXml(c);
         }
         return s;
      },
      escape: function(txt) {
         return txt.replace(/[\\]/g, "\\\\")
                   .replace(/[\"]/g, '\\"')
                   .replace(/[\n]/g, '\\n')
                   .replace(/[\r]/g, '\\r');
      },
      removeWhite: function(e) {
         e.normalize();
         for (var n = e.firstChild; n; ) {
            if (n.nodeType == 3) {  // text node
               if (!n.nodeValue.match(/[^ \f\n\r\t\v]/)) { // pure whitespace text node
                  var nxt = n.nextSibling;
                  e.removeChild(n);
                  n = nxt;
               }
               else
                  n = n.nextSibling;
            }
            else if (n.nodeType == 1) {  // element node
               X.removeWhite(n);
               n = n.nextSibling;
            }
            else                      // any other node
               n = n.nextSibling;
         }
         return e;
      }
   };
   if (xml.nodeType == 9) // document node
      xml = xml.documentElement;
   var json = X.toJson(X.toObj(X.removeWhite(xml)), xml.nodeName, "\t");
   return "{\n\t" + ('\t' ? json.replace(/\t/g, '\t') : json.replace(/\t|\n/g, '')) + "\n}";
}
